versioned_docs/version-2020fa/lecture7.md

```diff
@@ -3,4 +3,412 @@ id: lecture7
+every time the Counter rerenders,
+
+```tsx
+useEffect(() => {
+  document.title = `You clicked ${count} times`;
+});
+```
+
+`useEffect` also has a optional second parameter, which is an array of dependencies.
+If any of these dependencies change (like state variables), the effect is run. For example,
+if we want to update the title of our webpage when the Counter from the previous example
+changes its `count` state, we would use the following `useEffect` Hook in the functional
+component:
+
+```tsx
+// This code will only call the function when the count changes - more optimized
+useEffect(() => {
+  document.title = `You clicked ${count} times`;
+}, [count]);
+```
+
+#### `useEffect` - Subscribing to Outside Data and Separating Concerns
+
+Oftentimes when developing web applications, we use APIs to continually update
+and serve data to the user. These can be our own homemade APIs through tools like
+Express, or external APIs. In any case, it can be wasteful to constantly subscribe
+to an API when the data isn't being rendered. In the second `useEffect` call below,
+assume we have a `ChatAPI` object we can subscribe to and receive data from, or
+unsubscribe to. We can return a **function** from a `useEffect` call to be run
+after it completes, and we pass it an arrow function that simply unsubscribes from the current API instance.
+
+Also, even though we might be using the same hook for managing side effects of
+different parts of a component, it's best practice to separate them into different
+function calls, as seen with the two `useEffect` functions below. Both of them will
+still be run when the component re-renders (e.x. when any state of the component or parent component is changed).
+
+```tsx title="FriendStatusWithCounter.tsx"
+type Props = { readonly friend: { readonly id: string } };
+
+const FriendStatusWithCounter = ({ friend }: Props) => {
+  const [count, setCount] = useState(0);
+  const [isOnline, setIsOnline] = useState(false);
+
+  useEffect(() => {
+    document.title = `You clicked ${count} times`;
+  });
+
+  useEffect(() => {
+    ChatAPI.subscribeToFriendStatus(friend.id, (status) => {
+      setIsOnline(status.online);
+    });
+    // Cleanup code
+    return () => ChatAPI.unsubscribeToFriendStatus(friend.id);
+  });
+
+  return (
+    <div>
+      Your friend with id {id} is {isOnline ? ‘Online’ : ‘Offline’}.
+    </div>
+  );
+};
+```
+
+### Important Note
+
+While searching online for help / debugging, you will often
+run into StackOverflow answers or other resources for React dating back from
+a few years ago. Unfortunately, when it comes to developing with modern web
+technologies like React, some solutions often include deprecated features of React
+in the present day or need to be translated to the new paradigm (like Hooks).
+It is up to the developer to find the best solution that compromises between
+the bleeding-edge and practicality.
```

Comment:
It would be cool to put this in a [callout/admonition](https://v2.docusaurus.io/docs/markdown-features#calloutsadmonitions) like caution?